### ISSUE TITLE
[HOLD for payment 2023-06-19] [$250] Migrate BigNumberPad.js to function component

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#16120]
### Filenames
- [ ] [src/components/BigNumberPad.js](https://github.com/Expensify/App/blob/main/src/components/BigNumberPad.js)
    - **type:** `React.PureComponent`
    - **has state values:** true
    - **has refs:** false
    - **has context:** false
    - **lifecycle methods:** none

### Task
- We currently have some class components in our codebase that we would like to refactor to a function component.
- Here's a link with some general advice on how to refactor a class component to a function component: https://react.dev/reference/react/Component#alternatives
- If you need additional guidance, please ask in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q)
- Test for any regressions and verify that there are no breaking changes

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c285211ad303e051</li>
        <li>Upwork Job ID: 1664696501929406464</li>
        <li>Last Price Increase: 2023-06-02</li>
    </ul>
</details>

**Comment by PauloGasparSv** [comment]
Still on hold, not overdue.

**Comment by PauloGasparSv** [comment]
This is still on HOLD so not overdue.

**Comment by PauloGasparSv** [comment]
No updates, still on hold

**Comment by PauloGasparSv** [comment]
No updates, still on hold

**Comment by PauloGasparSv** [comment]
No updates here

**Comment by PauloGasparSv** [comment]
No updates here

**Comment by marcaaron** [comment]
Heads up, I'm taking this issue off `HOLD` and making it a `Daily`. If you are unable to work on it please remove your assignment and add the `External` + `Help Wanted` labels as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1685727880152119).

**Comment by PauloGasparSv** [comment]
> If you are unable to work on it please remove your assignment and add the External + Help Wanted labels as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1685727880152119).

Will do, same as [here](https://github.com/Expensify/App/issues/16132#issuecomment-1574126826)

**Comment by mkhutornyi** [comment]
I can fix it immediately

**Comment by jayeshmangwani** [comment]
I would like to work on this

**Comment by mkhutornyi** [comment]
Contributor details
Your Expensify account email: mkhutorniy@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~010baa511a9de4eeec?viewMode=1

**Comment by Vishrut19** [comment]
Hey! This is my proposal for this job : 

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Problem : Migrate BigNumberPad.js to function component #16120

### What is the root cause of that problem?

The root cause of the problem is that most of the things written are in class based which needs to be converted into function based as the updated ReactJS syntax.

### What changes do you think we should make in order to solve the problem?

Before This is class based code:

```
class BigNumberPad extends React.PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            timer: null,
        };
    }

    /**
     * Handle long press key on number pad.
     * Only handles the '<' key and starts the continuous input timer.
     *
     * @param {String} key
     */
    handleLongPress(key) {
        // Only handles deleting.
        if (key !== '<') {
            return;
        }
        this.props.longPressHandlerStateChanged(true);
        const timer = setInterval(() => {
            this.props.numberPressed(key);
        }, 100);
        this.setState({timer});
    }

    render() {
        return (
            <View
                style={[styles.flexColumn, styles.w100]}
                nativeID={this.props.nativeID}
            >
                {_.map(padNumbers, (row, rowIndex) => (
                    <View
                        key={`NumberPadRow-${rowIndex}`}
                        style={[styles.flexRow, styles.mt3]}
                    >
                        {_.map(row, (column, columnIndex) => {
                            // Adding margin between buttons except first column to
                            // avoid unccessary space before the first column.
                            const marginLeft = columnIndex > 0 ? styles.ml3 : {};
                            return (
                                <Button
                                    key={column}
                                    shouldEnableHapticFeedback
                                    style={[styles.flex1, marginLeft]}
                                    text={column === '<' ? column : this.props.toLocaleDigit(column)}
                                    onLongPress={() => this.handleLongPress(column)}
                                    onPress={() => this.props.numberPressed(column)}
                                    onPressIn={ControlSelection.block}
                                    onPressOut={() => {
                                        clearInterval(this.state.timer);
                                        ControlSelection.unblock();
                                        this.props.longPressHandlerStateChanged(false);
                                    }}
                                    onMouseDown={(e) => e.preventDefault()}
                                />
                            );
                        })}
                    </View>
                ))}
            </View>
        );
    }
}
```

After (new function based code):

```

import React, { useState } from 'react';
import { View, Button } from 'react-native';
import _ from 'lodash';

const BigNumberPad = (props) => {
  const [timer, setTimer] = useState(null);

  /**
   * Handle long press key on number pad.
   * Only handles the '<' key and starts the continuous input timer.
   *
   * @param {String} key
   */
  const handleLongPress = (key) => {
    // Only handles deleting.
    if (key !== '<') {
      return;
    }
    props.longPressHandlerStateChanged(true);
    const newTimer = setInterval(() => {
      props.numberPressed(key);
    }, 100);
    setTimer(newTimer);
  };

  return (
    <View style={[styles.flexColumn, styles.w100]} nativeID={props.nativeID}>
      {_.map(padNumbers, (row, rowIndex) => (
        <View key={`NumberPadRow-${rowIndex}`} style={[styles.flexRow, styles.mt3]}>
          {_.map(row, (column, columnIndex) => {
            // Adding margin between buttons except first column to
            // avoid unnecessary space before the first column.
            const marginLeft = columnIndex > 0 ? styles.ml3 : {};
            return (
              <Button
                key={column}
                shouldEnableHapticFeedback
                style={[styles.flex1, marginLeft]}
                text={column === '<' ? column : props.toLocaleDigit(column)}
                onLongPress={() => handleLongPress(column)}
                onPress={() => props.numberPressed(column)}
                onPressIn={ControlSelection.block}
                onPressOut={() => {
                  clearInterval(timer);
                  ControlSelection.unblock();
                  props.longPressHandlerStateChanged(false);
                }}
                onMouseDown={(e) => e.preventDefault()}
              />
            );
          })}
        </View>
      ))}
    </View>
  );
};


```



### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by nhminhduc** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Problem : Migrate BigNumberPad.js to function component https://github.com/Expensify/App/issues/16120### What is the root cause of that problem?
### What changes do you think we should make in order to solve the problem?
New code.
```
import React, { useState } from 'react';
import { View } from 'react-native';
import _ from 'underscore';
import PropTypes from 'prop-types';
import styles from '../styles/styles';
import Button from './Button';
import ControlSelection from '../libs/ControlSelection';
import withLocalize, { withLocalizePropTypes } from './withLocalize';

const propTypes = {
    numberPressed: PropTypes.func.isRequired,
    longPressHandlerStateChanged: PropTypes.func,
    nativeID: PropTypes.string,
    ...withLocalizePropTypes,
};

const defaultProps = {
    longPressHandlerStateChanged: () => {},
    nativeID: 'numPadView',
};

const padNumbers = [
    ['1', '2', '3'],
    ['4', '5', '6'],
    ['7', '8', '9'],
    ['.', '0', '<'],
];

BigNumberPad.propTypes = propTypes;
BigNumberPad.defaultProps = defaultProps;

const BigNumberPad = ({
    numberPressed,
    longPressHandlerStateChanged,
    nativeID,
    toLocaleDigit,
}) => {
    const [timer, setTimer] = useState(null);

    const handleLongPress = (key) => {
        if (key !== '<') return;

        longPressHandlerStateChanged(true);
        const newTimer = setInterval(() => {
            numberPressed(key);
        }, 100);
        setTimer(newTimer);
    };

    return (
        <View
            style={[styles.flexColumn, styles.w100]}
            nativeID={nativeID}
        >
            {_.map(padNumbers, (row, rowIndex) => (
                <View
                    key={`NumberPadRow-${rowIndex}`}
                    style={[styles.flexRow, styles.mt3]}
                >
                    {_.map(row, (column, columnIndex) => {
                        const marginLeft = columnIndex > 0 ? styles.ml3 : {};
                        return (
                            <Button
                                key={column}
                                shouldEnableHapticFeedback
                                style={[styles.flex1, marginLeft]}
                                text={column === '<' ? column : toLocaleDigit(column)}
                                onLongPress={() => handleLongPress(column)}
                                onPress={() => numberPressed(column)}
                                onPressIn={ControlSelection.block}
                                onPressOut={() => {
                                    clearInterval(timer);
                                    ControlSelection.unblock();
                                    longPressHandlerStateChanged(false);
                                }}
                                onMouseDown={(e) => e.preventDefault()}
                            />
                        );
                    })}
                </View>
            ))}
        </View>
    );
};

export default withLocalize(BigNumberPad);
```
### What alternative solutions did you explore? (Optional)

**Comment by Nikhil-Vats** [comment]
Hi fellow contributors 👋, fyi - this is a migration issue and the first one who asks gets the job most likely! More info [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1685727880152119).

**Comment by Vishrut19** [comment]
> Hi fellow contributors 👋, fyi - this is a migration issue and the first one who asks gets the job most likely! More info [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1685727880152119).

Hey! I have submitted my proposal. Please accept it.

**Comment by Vishrut19** [comment]
I have applied on Upwork for this issue

**Comment by luacmartins** [comment]
As per [this comment](https://expensify.slack.com/archives/C01GTK53T8Q/p1685727880152119), no proposals is neceesary on this issue, so I'm assigning it to the first contributor @mkhutornyi

**Comment by mkhutornyi** [comment]
Issue fixing is no problem at all but I will create a PR for this in a day as I am setting up dev environment and trying to understand the project code base. 

I will apply for this job in upwork as soon as I create the PR..

**Comment by MahmudjonToraqulov** [comment]
I'd like to work on this issue

**Comment by ubaid-rehman** [comment]
Contributor details
Your Expensify account email: ubaid.dilawari@gmail.com 
Upwork Profile Link: https://www.upwork.com/freelancers/~013499d740f52794ab

**Comment by Basemsisy** [comment]
Hello There, I want to do this job, what should I do as the first step?

**Comment by Basemsisy** [comment]
Contributor details
Your Expensify account email: basemsisy@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01315cef840bf38fb3

**Comment by luacmartins** [comment]
@mkhutornyi did you manage to setup your environment?

**Comment by ubaid-rehman** [comment]
I wanted to start working on Expensify jobs from UpWork, I did the local setup but for Android and IOS their readme refers to a Stackoverflow teams account and that requires a @expensify email address to register. I was hoping that anyone would point me in the right direction as to how to get that or there is another way to access that.
Any help would be greatly appreciated.

**Comment by ubaid-rehman** [comment]
> You can read [this](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md). It explains everything.

@thesahindia thanks for the link but I have gone through this, may be I am missing something and couldn't find the SO account part in there. Could you please refer to the section / part where it is mentioned.

**Comment by ubaid-rehman** [comment]
> Ahh! My bad. I was in a hurry so I didn't really read your comment correctly.
> 
> > I did the local setup but for Android and IOS their readme refers to a Stackoverflow teams account and that requires a Expensify email address to register
> 
> Only Expensify employees can use the Expensify emails. So I would suggest you to join our slack channel. There you can find many threads related to the setup or you can just ask for help and others will guide you. You can send an email to [contributors@expensify.com](mailto:contributors@expensify.com) to get invited to the channel. Hope that helps.

Thanks again for the timely help. 
I have sent that email.

**Comment by mkhutornyi** [comment]
> @mkhutornyi did you manage to setup your environment?

Yes I managed to setup the environment and was able to run the app in all platforms.

**Comment by johncschuster** [comment]
Working on inviting the contributors to the Upwork job 👍

**Comment by johncschuster** [comment]
@mkhutornyi Can you let me know once you've applied to the job?

**Comment by johncschuster** [comment]
@mkhutornyi bump!

**Comment by luacmartins** [comment]
waiting on payment

**Comment by luacmartins** [comment]
Still waiting on payment. @mkhutornyi it seems lioke you need to apply to the job!

**Comment by johncschuster** [comment]
I've bumped @mkhutornyi in Slack 👍

**Comment by johncschuster** [comment]
I haven't heard from @mkhutornyi. 

@thesahindia, I've just issued payment to you. Sorry for bottlenecking payment while waiting on contact from @mkhutornyi. I didn't need to wait to issue your payment.

**Comment by luacmartins** [comment]
@johncschuster are we still waiting on @mkhutornyi?

**Comment by johncschuster** [comment]
Yep! We're waiting on @mkhutornyi to join the job so I can pay them on Upwork.

**Comment by johncschuster** [comment]
Bump!

**Comment by johncschuster** [comment]
@mkhutornyi bump!

**Comment by mkhutornyi** [comment]
Sorry for late. @johncschuster 
I was on vacation.
I will apply to job today.
Thanks.

**Comment by mkhutornyi** [comment]
Hi, @johncschuster 
I just send my proposal via upwork.
Thanks.

**Comment by johncschuster** [comment]
Thanks for sending that proposal, @mkhutornyi! I've just sent the offer. Please ping me when you've accepted so I can issue payment. Thank you!

**Comment by mkhutornyi** [comment]
> Thanks for sending that proposal, @mkhutornyi! I've just sent the offer. Please ping me when you've accepted so I can issue payment. Thank you!

Accepted your offer.
Thanks.

**Comment by johncschuster** [comment]
Payment has been issued! I _think_ we can close this one up. 🎉
